src/Instrumentation/Psr15/src/Psr15Instrumentation.php

```diff
@@ -0,0 +1,115 @@
+
+        /**
+         * Create a span for each psr-15 middleware that is executed.
+         */
+        hook(
+            MiddlewareInterface::class,
+            'process',
+            static function (MiddlewareInterface $middleware, array $params, string $class, string $function, ?string $filename, ?int $lineno) use ($instrumentation) {
+                $span = $instrumentation->tracer()->spanBuilder(sprintf('%s::%s', $class, $function))
+                    ->setAttribute('code.function', $function)
+                    ->setAttribute('code.namespace', $class)
+                    ->setAttribute('code.filepath', $filename)
+                    ->setAttribute('code.lineno', $lineno)
+                    ->startSpan();
+
+                Context::storage()->attach($span->storeInContext(Context::getCurrent()));
+            },
+            static function (MiddlewareInterface $middleware, array $params, ?ResponseInterface $response, ?Throwable $exception) {
+                $scope = Context::storage()->scope();
+                if (!$scope) {
+                    return;
+                }
+                $scope->detach();
+                $span = Span::fromContext($scope->context());
+                if ($exception) {
+                    $span->recordException($exception, [TraceAttributes::EXCEPTION_ESCAPED => true]);
+                    $span->setStatus(StatusCode::STATUS_ERROR, $exception->getMessage());
+                }
+                $span->end();
+            }
+        );
+
+        /**
+         * Create a span to wrap RequestHandlerInterface::handle
+         * Stores the span as a request attribute which may be accessed by later hooks.
+         */
+        hook(
+            RequestHandlerInterface::class,
+            'handle',
+            static function (RequestHandlerInterface $handler, array $params, string $class, string $function, ?string $filename, ?int $lineno) use ($instrumentation) {
+                $request = ($params[0] instanceof ServerRequestInterface) ? $params[0] : null;
+                $parent = Context::getCurrent();
+                if ($request instanceof ServerRequestInterface && !$request->getAttribute(SpanInterface::class)) {
+                    //extract from request headers
+                    $parent = Globals::propagator()->extract($request->getHeaders());
+                }
+                $span = $instrumentation->tracer()->spanBuilder(sprintf('HTTP %s', $request?->getMethod() ?? 'unknown'))
+                    ->setParent($parent)
+                    ->setSpanKind(SpanKind::KIND_SERVER)
+                    ->setAttribute('code.function', $function)
+                    ->setAttribute('code.namespace', $class)
+                    ->setAttribute('code.filepath', $filename)
+                    ->setAttribute('code.lineno', $lineno)
+                    ->setAttribute(TraceAttributes::HTTP_URL, $request?->getUri()->__toString())
```

Comment:
Does it make sense to set these attributes in case request is null?

Reply:
I think it does...otherwise it would be more difficult to identify where the span came from, if you were looking at a mostly-empty span in a trace.